apiserver/tools.go

```diff
@@ -20,20 +20,121 @@ import (
+	// The following errors can be returned:
+	// - [coreerrors.NotSupported] if the architecture is not supported.
+	// - [agentbinaryerrors.AlreadyExists] if an agent binary already exists for
+	// this version and architecture.
+	// - [agentbinaryerrors.ObjectNotFound] if there was a problem referencing
+	// the agent binary metadata with the previously saved binary object.
+	// This error should be considered an internal problem. It is discussed here
+	// to make the caller aware of future problems.
+	// - [coreerrors.NotValid] if the agent version is not valid or the SHA256
+	// hash doesn't match the generated hash.
+	AddAgentBinaryWithSHA256(
+		context.Context,
+		io.Reader,
+		coreagentbinary.Version,
+		int64,
+		string,
+	) error
+}
+
+// AgentBinaryStoreGetter is a deferred type that can be used to get an
+// AgentBinaryStore at exactly the time it is needed. This allows for the
+// context aware answers to be made.
+type AgentBinaryStoreGetter func(*http.Request) (AgentBinaryStore, error)
+
+type BlockChecker interface {
+	// ChangeAllowed checks if change block is in place.
+	// Change block prevents all operations that may change
+	// current model in any way from running successfully.
+	ChangeAllowed(context.Context) error
+}
+
+// DomainServicesGetter describes a type that can be used for getting
+// [services.DomainServices] from a given context that comes from a http
+// request.
+type DomainServicesGetter func(ctx context.Context) (services.DomainServices, error)
+
+// BlockCheckerGetterForServices returns a [BlockCheckerGetter] that is
+// constructed from the supplied context.
+func BlockCheckerGetterForServices(servicesGetter DomainServicesGetter) func(context.Context) (BlockChecker, error) {
+	return func(ctx context.Context) (BlockChecker, error) {
+		svc, err := servicesGetter(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		return common.NewBlockChecker(svc.BlockCommand()), nil
+	}
+}
+
+// controllerAgentBinaryStoreForHTTPContext provides a deferred getter that
+// will provide the controllers [AgentBinaryStore] for the given [httpContext].
+func controllerAgentBinaryStoreForHTTPContext(httpCtx httpContext) AgentBinaryStoreGetter {
+	return func(r *http.Request) (AgentBinaryStore, error) {
+		services, err := httpCtx.domainServicesForRequest(r.Context())
+		if err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+
+		return services.AgentBinaryStore(), nil
+	}
+}
+
+// migratingAgentBinaryStoreForHTTPContext provides a deferred getter that will
+// provide the agent binary store for the model that is being migrated as part
+// of the request.
+func migratingAgentBinaryStoreForHTTPContext(httpCtx httpContext) AgentBinaryStoreGetter {
+	return func(r *http.Request) (AgentBinaryStore, error) {
+		services, err := httpCtx.domainServicesDuringMigrationForRequest(r)
+		if err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+
+		// TODO (tlm): Add model binary store here.
```

Comment:
To confirm - the intent is to add a `ModelAgentBinaryStore() *agentbinaryservice.AgentBinaryStore` method to the `ModelDomainServices` interface?

Reply:
I was waiting for the other PR to land. I will fix this up now and remove the todo.

Reply:
Done & Done